controllers/grafananotificationtemplate_controller.go

```diff
@@ -0,0 +1,213 @@
+	return ctrl.Result{RequeueAfter: notificationTemplate.Spec.ResyncPeriod.Duration}, nil
+}
+
+func (r *GrafanaNotificationTemplateReconciler) reconcileWithInstance(ctx context.Context, instance *grafanav1beta1.Grafana, notificationTemplate *grafanav1beta1.GrafanaNotificationTemplate) error {
+	cl, err := client2.NewGeneratedGrafanaClient(ctx, r.Client, instance)
+	if err != nil {
+		return fmt.Errorf("building grafana client: %w", err)
+	}
+
+	trueRef := "true" //nolint:goconst
+	editable := true
+	if notificationTemplate.Spec.Editable != nil && !*notificationTemplate.Spec.Editable {
+		editable = false
+	}
+
+	var updatedNT models.NotificationTemplateContent
+	updatedNT.Template = notificationTemplate.Spec.Template
+	params := provisioning.NewPutTemplateParams().WithName(notificationTemplate.Spec.Name).WithBody(&updatedNT)
+	if editable {
+		params.SetXDisableProvenance(&trueRef)
+	}
+	_, err = cl.Provisioning.PutTemplate(params) //nolint:errcheck
+	if err != nil {
+		return fmt.Errorf("creating or updating notification template: %w", err)
+	}
+	return nil
+}
+
+func (r *GrafanaNotificationTemplateReconciler) getNotificationTemplateByName(ctx context.Context, instance *grafanav1beta1.Grafana, notificationTemplate *grafanav1beta1.GrafanaNotificationTemplate) (models.NotificationTemplate, error) {
+	cl, err := client2.NewGeneratedGrafanaClient(ctx, r.Client, instance)
+	if err != nil {
+		return models.NotificationTemplate{}, fmt.Errorf("building grafana client: %w", err)
+	}
+
+	params := provisioning.NewGetTemplateParams().WithName(notificationTemplate.Spec.Name)
+	remote, err := cl.Provisioning.GetTemplateWithParams(params)
+	if err != nil {
+		return models.NotificationTemplate{}, fmt.Errorf("getting notification template: %w", err)
+	}
+	return *remote.Payload, nil
+}
+
+func (r *GrafanaNotificationTemplateReconciler) finalize(ctx context.Context, notificationTemplate *grafanav1beta1.GrafanaNotificationTemplate) error {
+	r.Log.Info("Finalizing GrafanaNotificationTemplate")
+
+	instances, err := GetAllMatchingInstances(ctx, r.Client, notificationTemplate)
+	if err != nil {
+		return fmt.Errorf("fetching instances: %w", err)
+	}
+	for _, i := range instances {
+		instance := i
+		if err := r.removeFromInstance(ctx, &instance, notificationTemplate); err != nil {
+			return fmt.Errorf("removing contact point from instance: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (r *GrafanaNotificationTemplateReconciler) removeFromInstance(ctx context.Context, instance *grafanav1beta1.Grafana, notificationTemplate *grafanav1beta1.GrafanaNotificationTemplate) error {
+	cl, err := client2.NewGeneratedGrafanaClient(ctx, r.Client, instance)
+	if err != nil {
+		return fmt.Errorf("building grafana client: %w", err)
+	}
+
+	_, err = r.getNotificationTemplateByName(ctx, instance, notificationTemplate)
+	if err != nil {
+		return fmt.Errorf("getting notification template by name: %w", err)
+	}
```

Comment:
```suggestion
```

If a template is absent, it should not be considered an error, otherwise we'll have a never-ending reconciliation. Grafana API returns 204 (No Content) when someone attempts to remove non-existing template, it's treated as success by the Grafana API client we're using.